### PR TITLE
[FLINK-31444][runtime]FineGrainedSlotManager reclaims slots when job is finished

### PR DESCRIPTION
**Pull request by huwh**
## What is the purpose of the change
When a job shuts down there is a race condition between slots being freed and requirements being set to 0. If the slot release arrives first at the RM then it will immediately try to re-allocate slots, since the requirements are not 0 yet.

This change try to free all inactive slots in task executor when job finished(clear requirements). 

## Verifying this change
  - *Added unit test testReclaimInactiveSlotsOnClearRequirements to verify reclaim only invoked on clear requirements*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation
  - Does this pull request introduce a new feature?(no)

### CODE REVIEW

**Review by reswqa — COMMENTED** [review]
Thanks @huwh for porting this feature to `FineGrainedSlotManager`, this change looks good to me overall, I only left a minor comments, PTAL.

File: flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultSlotStatusSyncer.java [hunk 1]
```diff
@@ -308,6 +309,14 @@ private boolean syncAllocatedSlotStatus(SlotStatus slotStatus, TaskManagerInfo t
         }
     }
 
+    @Override
+    public void freeInactiveSlots(JobID jobId, TaskExecutorConnection taskExecutorConnection) {
+        checkStarted();
+        taskExecutorConnection
+                .getTaskExecutorGateway()
+                .freeInactiveSlots(jobId, taskManagerRequestTimeout);
+    }
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultSlotStatusSyncer.java:318]: Shall we eagerly notify the task manager tracker and resource tracker about the freeing of the slots? Or we rely on task managers to report the slots becoming available later to keep the states consistent? I personally lean towards eagerly sync the status. Either way, it should be documented.

**Review by xintongsong — CHANGES_REQUESTED** [review]

**Comment by huwh** [comment]
@xintongsong thanks, comments addressed. PTAL

**Comment by xintongsong** [comment]
@flinkbot run azure

**Review by xintongsong — APPROVED** [review]
LGTM. Merging this.

**Pull request closed without merging**